akka-stream/src/main/scala/akka/stream/impl/BoundedQueueSource.scala

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2020-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import java.util.concurrent.atomic.AtomicReference
+
+import scala.annotation.tailrec
+
+import akka.annotation.InternalApi
+import akka.dispatch.AbstractBoundedNodeQueue
+import akka.stream._
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, OutHandler, StageLogging }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object BoundedQueueSourceStage {
+  sealed trait State
+  case object NeedsActivation extends State
+  case object Running extends State
+  case class Done(result: QueueCompletionResult) extends State
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final class BoundedQueueSourceStage[T](bufferSize: Int)
+    extends GraphStageWithMaterializedValue[SourceShape[T], BoundedQueueSource[T]] {
+  import BoundedQueueSourceStage._
+
+  require(bufferSize >= 0, "BoundedQueueSourceStage.bufferSize must be >= 0")
```

Comment:
As noted above, does it really make sense that it can be 0?

Reply:
Updated to be `> 0`